cell2location/distributions/AutoAmortisedNormalMessenger.py

```diff
@@ -0,0 +1,485 @@
+
+    def median(self, *args, **kwargs):
+        self._computing_median = True
+        try:
+            return self(*args, **kwargs)
+        finally:
+            self._computing_median = False
+
+    @torch.no_grad()
+    def _get_posterior_median(self, name, prior):
+        transform = biject_to(prior.support)
+        if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
+            loc, scale, weight = self._get_params(name, prior)
+            loc = loc + transform.inv(prior.mean) * weight
+        else:
+            loc, scale = self._get_params(name, prior)
+        return transform(loc)
+
+    def quantiles(self, quantiles, *args, **kwargs):
+        self._computing_quantiles = True
+        self._quantile_values = quantiles
+        try:
+            return self(*args, **kwargs)
+        finally:
+            self._computing_quantiles = False
+
+    @torch.no_grad()
+    def _get_posterior_quantiles(self, name, prior):
+        transform = biject_to(prior.support)
+        if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
+            loc, scale, weight = self._get_params(name, prior)
+            loc = loc + transform.inv(prior.mean) * weight
+        else:
+            loc, scale = self._get_params(name, prior)
+
+        site_quantiles = torch.tensor(self._quantile_values, dtype=loc.dtype, device=loc.device)
+        site_quantiles_values = dist.Normal(loc, scale).icdf(site_quantiles)
+        return transform(site_quantiles_values)
+
+    def mutual_information(self, *args, **kwargs):
+        # compute samples necessary to compute MI
+        self.samples_for_mi = self(*args, **kwargs)
+        self._computing_mi = True
+        try:
+            # compute mi (saved to self.mi)
+            self(*args, **kwargs)
+            return self.mi
+        finally:
+            self._computing_mi = False
+
+    @torch.no_grad()
+    def _get_mutual_information(self, name, prior):
+        """Approximate the mutual information between data x and latent variable z
+
+            I(x, z) = E_xE_{q(z|x)}log(q(z|x)) - E_xE_{q(z|x)}log(q(z))
+
+        Returns: Float
+
+        """
+
+        #### get posterior mean and variance ####
+        transform = biject_to(prior.support)
+        if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
+            loc, scale, weight = self._get_params(name, prior)
+            loc = loc + transform.inv(prior.mean) * weight
+        else:
+            loc, scale = self._get_params(name, prior)
+
+        if name not in self.amortised_plate_sites["sites"].keys():
+            # if amortisation is not used for a particular site return MI=0
+            return 0
+
+        #### create tensors with useful numbers ####
+        one = torch.ones((), dtype=loc.dtype, device=loc.device)
+        two = torch.tensor(2, dtype=loc.dtype, device=loc.device)
+        pi = torch.tensor(3.14159265359, dtype=loc.dtype, device=loc.device)
```

Comment:
I think you can just use `1`, `2`, and `torch.pi` here
Also why is `one` set to `()`?

Reply:
All 3 should be scalar tensors, have the same dtype and device as loc. For some reason, when using torch.pi, I got errors with device location.